tests/modules/layers/test_anyprecision_optimizer.py

```diff
@@ -0,0 +1,98 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from copy import deepcopy
+
+import pytest
+import torch
+import torch.nn as nn
+import torch.optim as optim
+
+from tests.test_utils import assert_expected, gpu_test
+from torchmultimodal.modules.optimizers.anyprecision import AnyPrecisionAdamW
+
+
+class TestAnyPrecisionOptimizer:
+    def _test_adam_equivalence(self, model, model_clone):
+        # Test non-default options
+        betas = (0.8, 0.88)
+        weight_decay = 0.03
+
+        adam_opt = optim.AdamW(
+            model_clone.parameters(), betas=betas, weight_decay=weight_decay
+        )
+        anyprecision_adam = AnyPrecisionAdamW(
+            model.parameters(),
+            variance_dtype=torch.float32,
+            betas=betas,
+            weight_decay=weight_decay,
+        )
+
+        # Verify params are equal initially
+        model_orig_params = [p.clone() for p in model.parameters()]
+        for p1, p2 in zip(model_clone.parameters(), model_orig_params):
+            assert_expected(p1, p2)
+
+        for i in range(6):
+            adam_opt.zero_grad()
+            anyprecision_adam.zero_grad()
+            inp = torch.randn(5, 5, device=next(model.parameters()).device)
+            model(inp).sum().backward()
+            model_clone(inp).sum().backward()
+            adam_opt.step()
+            anyprecision_adam.step()
+
+            # Ensure params are modified from original
+            if i == 0:
+                for p1, p2 in zip(model.parameters(), model_orig_params):
+                    assert not torch.equal(p1, p2)
+
+            for p1, p2 in zip(model.parameters(), model_clone.parameters()):
+                assert_expected(p1, p2)
+
+    def test_adam_equivalence(self, device="cpu"):
+        """
+        Tests that AnyPrecisionAdamW is equivalent to AdamW when
+        kahan summation and different dtypes for momentum, variance,
+        and compensation buffer are turned off (i.e. all float32).
+        """
+        if device == "cuda" and not torch.cuda.is_available():
```

Comment:
Sorry maybe I'm missing something: when would device ever not be cpu in the current setup?

Reply:
I was hoping gpu would be added based on triton kernels being included (which requires gpu). 
I was also hitting an issue running on cpu due to an upcast error. 
I've reworked the tests to now function on cpu and avoid the upcasting issue, and opened this issue in PT to get the cpu upcast issue resolved for the future:
https://github.com/pytorch/pytorch/issues/110905

I've thus removed the gpu_test from this one and the new kahan_summation test, so we are now 100% running on cpu. 
<img width="1041" alt="anyPrecision_3Pass" src="https://github.com/facebookresearch/multimodal/assets/46302957/571e7264-6575-48fb-90b6-d5a16811efa7">